src/v/cluster/partition_balancer_planner.cc

```diff
@@ -682,26 +707,28 @@ auto partition_balancer_planner::request_context::do_with_partition(
     return visitor(part);
 }
 
-void partition_balancer_planner::request_context::for_each_partition(
+ss::future<> partition_balancer_planner::request_context::for_each_partition(
   ss::noncopyable_function<ss::stop_iteration(partition&)> visitor) {
-    for (const auto& t : _parent._state.topics().topics_map()) {
-        for (const auto& a : t.second.get_assignments()) {
-            auto ntp = model::ntp(t.first.ns, t.first.tp, a.id);
+    const auto& topics = _parent._state.topics();
+    for (auto it = topics_begin(topics); it != topics_end(topics); ++it) {
+        for (const auto& a : it->second.get_assignments()) {
+            auto ntp = model::ntp(it->first.ns, it->first.tp, a.id);
             auto stop = do_with_partition(ntp, a.replicas, visitor);
             if (stop == ss::stop_iteration::yes) {
-                return;
+                co_return;
             }
         }
+        co_await maybe_yield();
     }
 }
 
-void partition_balancer_planner::request_context::with_partition(
+ss::future<> partition_balancer_planner::request_context::with_partition(
```

Comment:
Should this be a coroutine, if we only do co_returns? Though it is used in a few lengthy loops so maybe it should yield too?

Reply:
making it a coroutine makes the caller code simple as co_awaiting on it means that we don't need to yield separately.